pkg/sql/catalog/descs/collection.go

```diff
@@ -264,12 +267,90 @@ func (tc *Collection) WriteDescToBatch(
 	descKey := catalogkeys.MakeDescMetadataKey(tc.codec(), desc.GetID())
 	proto := desc.DescriptorProto()
 	if kvTrace {
-		log.VEventf(ctx, 2, "Put %s -> %s", descKey, proto)
+		log.VEventf(ctx, 2, "CPut %s -> %s", descKey, proto)
 	}
 	b.CPut(descKey, proto, expected)
 	return nil
 }
 
+// DeleteDescToBatch adds a delete from system.descriptor to the batch.
+func (tc *Collection) DeleteDescToBatch(
+	ctx context.Context, kvTrace bool, id descpb.ID, b *kv.Batch,
+) error {
+	if id == descpb.InvalidID {
+		return errors.AssertionFailedf("cannot delete descriptor with an empty ID: %v", id)
+	}
+	descKey := catalogkeys.MakeDescMetadataKey(tc.codec(), id)
+	if kvTrace {
+		log.VEventf(ctx, 2, "Del %s", descKey)
+	}
+	b.Del(descKey)
+	tc.NotifyOfDeletedDescriptor(id)
+	return nil
+}
+
+// InsertNamespaceEntryToBatch adds an insertion into system.namespace to the
+// batch.
+func (tc *Collection) InsertNamespaceEntryToBatch(
+	ctx context.Context, kvTrace bool, e catalog.NameEntry, b *kv.Batch,
+) error {
+	if cachedID := tc.stored.GetCachedIDByName(e); cachedID != descpb.InvalidID {
```

Comment:
Why do we want to modify the the `stored` layer which is supposed to mirror the committed storage data? For read path (name resolution) we have `uncommitted descriptor` layer to read from uncommitted cache.

Same question for `upsert` and `delete` below.

Reply:
This is a good question. The short answer is that the name changes in the uncommitted layer must "shadow" the stored layer somehow. The stored layer has a cache, for better or for worse; if we don't evict entries from that cache we end up looking up stale names. Schemas in particular are susceptible to this because we look for their name -> ID mappings in the parent database descriptor if we have it...

Also, notice that we don't actually evict them completely, only from the by-name-index.

I think the a better solution might be to further dissociate the stored layer from its cache. They're so tightly coupled mostly for historical reasons. I'll think about this some more.

Reply:
Ah yeah, good point. uncommitted layer doesn't have the stale name.